packages/pyright-internal/src/languageService/completionProvider.ts

```diff
@@ -1427,6 +1456,299 @@ export class CompletionProvider {
+                            if (excludes.has(key)) {
+                                return;
+                            }
+                            excludes.add(key);
+
+                            this._addStringLiteralToCompletionList(
+                                key,
+                                quoteValue ? quoteValue.stringValue : undefined,
+                                postText,
+                                quoteValue
+                                    ? quoteValue.quoteCharacter
+                                    : this._parseResults.tokenizerOutput.predominantSingleQuoteCharacter,
+                                completionList
+                            );
+                        }
+                    );
+                });
+            }
+        }
+    }
+
+    private _getTypedDicts(baseExpressionNode: ExpressionNode, parseNode: ParseNode, type: ClassType | UnionType) {
+        if (!(tokenIsAtKey(this) || this._isInDictionaryStringKey(parseNode))) {
+            return [];
+        }
+
+        // path to resolve nested dictionaries
+        let path: DictPathEntry[] = [];
+        let curNode: ParseNode | undefined = parseNode;
+
+        while (curNode && curNode.id !== baseExpressionNode.id) {
+            // List is here as it is a supported container type
+            if (
+                curNode.nodeType === ParseNodeType.Dictionary ||
+                curNode.nodeType === ParseNodeType.Set ||
+                curNode.nodeType === ParseNodeType.List
+            ) {
+                if (
+                    curNode.parent?.nodeType === ParseNodeType.DictionaryKeyEntry &&
+                    curNode.parent?.keyExpression.nodeType === ParseNodeType.StringList
+                ) {
+                    path.push({
+                        entry: curNode.parent,
+                        key: curNode.parent.keyExpression.strings.map((s) => s.value).join(''),
+                    });
+                }
+            }
+
+            curNode = curNode.parent;
+        }
+
+        path = path.reverse();
+        const types = this._getTypedDictsRecursive(baseExpressionNode, parseNode, type, path);
+        const keys = this._getDictExpressionStringKeys(parseNode, new Set([parseNode.parent?.id]));
+        return this._tryNarrowTypedDicts(types, keys);
+
+        function tokenIsAtKey(provider: CompletionProvider): boolean {
+            // this function is only designed return true for states that
+            // are discarded by the parse tree, for example:
+            // {'foo': 'bar', '|}
+            const offset = convertPositionToOffset(provider._position, provider._parseResults.tokenizerOutput.lines)!;
+            const tokens = provider._parseResults.tokenizerOutput.tokens;
+            const index = tokens.getItemAtPosition(offset);
+            const token = tokens.getItemAt(index);
+            const prevToken = tokens.getItemAt(index - 1);
```

Comment:
what is index == 0, I think this might be good to be in like parsetreeutils.ts file so can be reused.

getTokenAt(parseResults, position) and getPreviousToken(parseResults, token) ?

Reply:
index === 0 is the current position of the cursor. Yeah that's a good suggestion.

Reply:
sorry, typo, what I meant was what "if" index == 0 :)

Reply:
Ah I see, that would be a bug, thanks for spotting it.

I don't think the index could be 0 in this context but it would probably be better to safeguard against it anyway.